File: cses/04_graph_algorithms/1693_teleporters-path/solution.py
import sys


def main():
    # Đọc toàn bộ input một lần để tránh chậm do I/O trên CPython.
    data = sys.stdin.buffer.read().split()
    idx = 0
    n = int(data[idx]); idx += 1
    m = int(data[idx]); idx += 1

    # Xây adjacency list cho đồ thị có hướng và đếm bậc vào/ra mỗi đỉnh.
    adj = [[] for _ in range(n + 1)]
    outdeg = [0] * (n + 1)
    indeg = [0] * (n + 1)

    for _ in range(m):
        a = int(data[idx]); idx += 1
        b = int(data[idx]); idx += 1
        adj[a].append(b)
        outdeg[a] += 1
        indeg[b] += 1

    # Điều kiện bậc cho đường đi Euler có hướng từ đỉnh 1 tới đỉnh n:
    #   đỉnh 1 dư đúng một cạnh ra (+1), đỉnh n dư đúng một cạnh vào (-1),
    #   mọi đỉnh khác cân bằng bậc vào/ra (0). (n >= 2 nên 1 != n.)
    ok = True
    for v in range(1, n + 1):
        d = outdeg[v] - indeg[v]
        if v == 1:
            if d != 1:
                ok = False
                break
        elif v == n:
            if d != -1:
                ok = False
                break
        else:
            if d != 0:
                ok = False
                break

    if not ok:
        sys.stdout.write("IMPOSSIBLE\n")
        return

    # Hierholzer kiểu iterative xuất phát từ đỉnh 1; ptr[v] là con trỏ tới cạnh
    # chưa dùng kế tiếp của v nên mỗi cạnh chỉ được xét đúng một lần.
    ptr = [0] * (n + 1)
    stack = [1]
    path = []
    while stack:
        v = stack[-1]
        p = ptr[v]
        if p < len(adj[v]):
            # Còn cạnh chưa dùng: đi theo cạnh đó, đẩy đỉnh kề vào stack.
            ptr[v] = p + 1
            stack.append(adj[v][p])
        else:
            # Hết cạnh: pop đỉnh và thêm vào path (thứ tự ngược).
            path.append(v)
            stack.pop()

    # Phải dùng hết mọi cạnh => path có m + 1 đỉnh; đây cũng bao trọn điều kiện
    # liên thông. Nếu không đủ thì IMPOSSIBLE.
    if len(path) != m + 1:
        sys.stdout.write("IMPOSSIBLE\n")
        return

    # Đảo ngược để có thứ tự đi thực sự rồi ghi output một lần.
    path.reverse()
    sys.stdout.write(' '.join(map(str, path)))
    sys.stdout.write('\n')


main()

File: cses/04_graph_algorithms/1693_teleporters-path/solution.cpp
#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    // Đọc đồ thị có hướng: n đỉnh, m cạnh; đồng thời đếm bậc vào/ra mỗi đỉnh.
    int n, m;
    cin >> n >> m;
    vector<vector<int>> graph(n + 1);
    vector<int> inDegree(n + 1, 0);
    vector<int> outDegree(n + 1, 0);
    for (int edge = 0; edge < m; ++edge) {
        int from, to;
        cin >> from >> to;
        graph[from].push_back(to);
        ++outDegree[from];
        ++inDegree[to];
    }

    // Điều kiện bậc cho đường đi Euler có hướng từ đỉnh 1 tới đỉnh n:
    //   đỉnh 1 dư đúng một cạnh ra (+1), đỉnh n dư đúng một cạnh vào (-1),
    //   mọi đỉnh khác phải cân bằng bậc vào/ra (0). Vi phạm => IMPOSSIBLE.
    for (int vertex = 1; vertex <= n; ++vertex) {
        const int difference = outDegree[vertex] - inDegree[vertex];
        const int required = vertex == 1 ? 1 : (vertex == n ? -1 : 0);
        if (difference != required) {
            cout << "IMPOSSIBLE\n";
            return 0;
        }
    }

    // Hierholzer kiểu iterative xuất phát từ đỉnh 1; nextEdge[v] là con trỏ tới
    // cạnh chưa dùng kế tiếp của v nên mỗi cạnh chỉ được xét đúng một lần.
    vector<size_t> nextEdge(n + 1, 0);
    vector<int> stack{1};
    vector<int> path;
    path.reserve(m + 1);
    while (!stack.empty()) {
        const int vertex = stack.back();
        if (nextEdge[vertex] < graph[vertex].size()) {
            // Còn cạnh chưa dùng: đi theo cạnh đó, đẩy đỉnh kề vào stack.
            stack.push_back(graph[vertex][nextEdge[vertex]++]);
        } else {
            // Hết cạnh: pop đỉnh và thêm vào path (thứ tự ngược).
            path.push_back(vertex);
            stack.pop_back();
        }
    }

    // len(path) phải bằng m + 1 thì mới dùng hết mọi cạnh; đây cũng chính là
    // kiểm tra điều kiện liên thông.
    if (static_cast<int>(path.size()) != m + 1) {
        cout << "IMPOSSIBLE\n";
        return 0;
    }

    // Đảo ngược để có thứ tự đi thực sự; xác nhận đầu là 1 và cuối là n.
    reverse(path.begin(), path.end());
    if (path.front() != 1 || path.back() != n) {
        cout << "IMPOSSIBLE\n";
        return 0;
    }
    for (size_t i = 0; i < path.size(); ++i) {
        if (i > 0) {
            cout << ' ';
        }
        cout << path[i];
    }
    cout << '\n';
    return 0;
}
